variables.tf

```diff
@@ -14,168 +20,185 @@ variable "enhanced_error_handling" {
     ])
-    error_message = "`asn`: Allowed formats: `1-4294967295` or `1-65535.0-65535`."
+    error_message = "`asn`: Allowed formats are: `1-4294967295` or `1-65535.0-65535`."
   }
 
   validation {
     condition = alltrue([
-      for k, v in var.template_peer : can(regex("^.{0,254}$", v.description)) || v.description == null
+      for v in var.template_peers : can(regex("^.{0,254}$", v.description)) || v.description == null
     ])
     error_message = "`description`: Maximum characters: `254`."
   }
 
   validation {
     condition = alltrue([
-      for k, v in var.template_peer : try(contains(["fabric-internal", "fabric-external", "fabric-border-leaf"], v.peer_type), v.peer_type == null, false)
+      for v in var.template_peers : try(contains(["fabric-internal", "fabric-external", "fabric-border-leaf"], v.peer_type), v.peer_type == null)
     ])
     error_message = "`peer_type`: Allowed values are `fabric-internal`, `fabric-external` or `fabric-border-leaf`."
   }
 
   validation {
     condition = alltrue([
-      for k, v in var.template_peer : can(regex("^\\S*$", v.source_interface)) || v.source_interface == null
+      for v in var.template_peers : can(regex("^\\S*$", v.source_interface)) || v.source_interface == null
     ])
     error_message = "`source_interface`: Whitespaces are not allowed. Must match first field in the output of `show int brief`. Example: `eth1/1`."
   }
 
   validation {
     condition = alltrue(flatten([
-      for key, value in var.template_peer : value.address_family == null ? [true] : [
-        for k, v in value.address_family : contains(["ipv4_unicast", "ipv6_unicast", "l2vpn_evpn"], k)
+      for value in var.template_peers : value.address_families == null ? [true] : [
+        for v in value.address_families : contains(["ipv4_unicast", "ipv6_unicast", "l2vpn_evpn"], v.address_family)
       ]
     ]))
-    error_message = "`address_family`: Allowed map keys are `ipv4_unicast`, `ipv6_unicast` or `l2vpn_evpn`."
+    error_message = "`address_family`: Allowed values are `ipv4_unicast`, `ipv6_unicast` or `l2vpn_evpn`."
   }
 }
 
-variable "vrf" {
-  description = "BGP VRFs."
-  type = map(object({
+variable "vrfs" {
+  description = <<EOT
+  BGP VRF list.
+  List `neighbors`:
+  Allowed formats `neighbor`: `192.168.1.1` or `192.168.1.0/24`.
+  Choices `peer_type`: `fabric-internal`, `fabric-external`, `fabric-border-leaf`. Default value `peer_type`: `fabric-internal`.
+  List `address_families`:
+  Choices `address_family`: `ipv4_unicast`, `ipv6_unicast`, `l2vpn_evpn`.
+  EOT
+  type = list(object({
+    vrf                             = string
     router_id                       = optional(string)
     log_neighbor_changes            = optional(bool)
     graseful_restart_stalepath_time = optional(number)
     graseful_restart_restart_time   = optional(number)
-    neighbors = optional(map(object({
+    neighbors = optional(list(object({
+      neighbor         = string
```

Comment:
Would `ip` be more descriptive?

Reply:
The logic that I tried to follow is to use plural for the list name and singular for the list element "key". For example:
- `template_peers` -> `template_peer`
- `vrfs` -> `vrf`
- `address_families` -> `address_family`
- `neighbors` -> `neighbor`

```
variable "vrfs" {
  type = list(object({
    vrf                             = string <---
    neighbors = optional(list(object({ <---
      neighbor         = string <---
      address_families = optional(list(object({ <---
        address_family          = string <---
        
variable "template_peers" {  <---
  type = list(object({
    template_peer    = string  <---
    address_families = optional(list(object({  <---
      address_family          = string  <---
```

Reply:
I understand, but the key does not necessarily need to be named after the list. Also, in case the list has more than one key attribute, this logic would no longer work. I would rather use a descriptive name like "ip" for neighbor or "name" for template.

Reply:
Agreed, let's do it this way.

Reply:
fixed